### ISSUE TITLE
[Form] Add "choice_labels", "choice_values", "choice_attr" and "group_by" options to ChoiceType

### ISSUE THREAD

**Issue by webmozart** [symfony/symfony#4067]
Depends on: #3879, #3968 (`OptionsParser::replaceDefaults` is required)

I want to add three options to the ChoiceType that abstract functionality of EntityType for use in all choice fields. See also #3479.

The new options are:

##### "choice_labels"

If given, the "choices" option is interpreted as storing the choices in the values instead of the keys. This way, choices can be created that select between values that are not allowed to be passed as array keys (any non-integer or non-string value).

"choice_labels" can be:

* an array with a label for each choice
* a callable that is executed for each choice and returns the label
* a property path string, if the choices are objects

```php
<?php
// The following two are equivalent
$builder->add('foo', 'choice', array(
    'choices' => array(0 => 'no', 1 => 'yes'),
));
$builder->add('foo', 'choice', array(
    'choices' => array(0, 1),
    'choice_labels' => array('no', 'yes'),
));

// The following ones cannot be expressed without the new option
$builder->add('foo', 'choice', array(
    'choices' => array(null, false, true),
    'choice_labels' => array('maybe', 'no', 'yes'),
));
$builder->add('foo', 'choice', array(
    'choices' => array($obj1, $obj2),
    'choice_labels' => 'name',
));
$builder->add('foo', 'choice', array(
    'choices' => array($obj1, $obj2),
    'choice_labels' => function ($obj) {
        return 'Label for ' . $obj->name;
    },
));
```

##### "choice_values"

Like "choice_labels", but stores/generates the values that are stored in the "value" attributes of the generated HTML. Values  must be strings and unique (i.e. no two choices must have the same value).

By default, the choices are used as values, if possible. If not possible, integer values are generated.

Like "choice_labels", this option can be an array, a callable or a string (if choices are objects).

```php
<?php
$builder->add('foo', 'choice', array(
    'choices' => array($obj1, $obj2),
    'choice_labels' => 'name',
    'choice_values' => 'id',
));
```

##### "choice_attr"

Like choice_values.

##### "group_by"

The "group_by" option stores a callable that returns the choice group for each choice or a property path if the choices are objects.

```php
<?php
$builder->add('favNumber', 'choice', array(
    'choices' => range(0, 100),
    'choice_labels' => range(0, 100),
    'group_by' => function ($choice) {
        return $choice % 2 == 0 ? 'Even' : 'Odd',
    }
));
```

##### "preferred_choices"

The "preferred_choices" option allows to pass the values of the preferred choices as array, a callable or a property path if the choices are objects.

```php
$builder->add('foo', 'choice', array(
    'choices' => array(null, false, true),
    'choice_labels' => array('maybe', 'no', 'yes'),
    'preferred_choices' => array(true, false),
));
$builder->add('foo', 'choice', array(
    'choices' => array($obj1, $obj2),
    'preferred_choices' => 'preferred', // i.e. isPreferred()
));
$builder->add('foo', 'choice', array(
    'choices' => array($obj1, $obj2),
    'preferred_choices' => function ($obj) {
        return $obj instanceof PreferredSomething;
    },
));
```

#### Poll

If the "choices" option is given as hierarchical array sorting each choice into a choice group, there are two possible ways for associating choices with their labels/values. Which one do you prefer?

##### A. Duplicate the hierarchy

```php
<?php
$builder->add('favNumber', 'choice', array(
    'choices' => array(
        'Even' => array($obj0, $obj2, $obj4),
        'Odd' => array($obj1, $obj3, $obj5),
    ),
    'choice_labels' => array(
        'Even' => array('Zero', 'Two', 'Four'),
        'Odd' => array('One', 'Three', 'Five'),
    ),
    'choice_values' => array(
        'Even' => array(0, 2, 4),
        'Odd' => array(1, 3, 5),
    ),
));
```

##### B. Array-index based

```php
<?php
$builder->add('favNumber', 'choice', array(
    'choices' => array(
        'Even' => array(0 => $obj0, 2 => $obj2, 4 => $obj4),
        'Odd' => array(1 => $obj1, 3 => $obj3, 5 => $obj5),
    ),
    'choice_labels' => array(
       array(0 => 'Zero', 1 => 'One', 2 => 'Two', 3 => 'Three', 4 => 'Four', 5 => 'Five'),
    ),
    'choice_values' => array(
        array(0 => 0, 1 => 1, 2 => 2, 3 => 3, 4 => 4, 5 => 5),
    ),
));
```

Please name your preferred option in the comments.

**Comment by mvrhov** [comment]
@webmozart: Any chances of getting this PR and the ones depending on it through so they can be in 2.7?

**Comment by webmozart** [comment]
Working on it.

**Comment by Kharestani** [comment]
Thanks for your awesomeness!
I propose you add new option "choice_labels_attr" to change choice labels attributes dynamically!
I know it might be too far but I think everything is too dynamic now and it's needed!

Thanks again.

**Comment by xabbuh** [comment]
@Kharestani Better create a new issue if you do not want your feature request to get lost (but please check first that there isn't already a similar issue).

**Comment by HeahDude** [comment]
see #16834

**Comment by Kharestani** [comment]
@HeahDude
My Bad!
Thanks!

**Comment by ABM-Dan** [comment]
What happened to choice_attr?

**Comment by xabbuh** [comment]
@ABM-Dan What do you mean?

**Comment by ABM-Dan** [comment]
@xabbuh I couldn't find any documentation for EntityType's choice_attr. After some investigating, it turns out it's only in the docs for 2.7, with an example that suits ChoiceType, but is a bit lackluster when it comes to EntityType (for instance, mentioning that $val in the example is an actual Entity, is definetly relevant).

**Comment by xabbuh** [comment]
@ABM-Dan Can you please open a new issue on the docs repository if you think we can improve the documentation somehow?

**Comment by ABM-Dan** [comment]
Will do at a later date.

**Comment by HeahDude** [comment]
@xabbuh what about symfony/symfony-docs#6260 ?

@ABM-Dan be careful if you preset `null` as default data, it will throw an exception in you type hint the callable see https://github.com/symfony/symfony/issues/17736#issuecomment-182390504.

I've planned to open an issue in the docs for it as type hinting in the best practices, but haven't done it yet.

**Comment by mandelaloko** [comment]
Nice!
